protocol/x/vault/keeper/shares.go

```diff
@@ -47,3 +56,72 @@ func (k Keeper) getTotalSharesIterator(ctx sdk.Context) storetypes.Iterator {
 
 	return storetypes.KVStorePrefixIterator(store, []byte{})
 }
+
+// MintShares mints shares of a vault for `owner` based on `quantumsToDeposit` by:
+// 1. Increasing total shares of the vault.
+// 2. Increasing owner shares of the vault for given `owner`.
+func (k Keeper) MintShares(
+	ctx sdk.Context,
+	vaultId types.VaultId,
+	owner string,
+	quantumsToDeposit *big.Int,
+) error {
+	// Quantums to deposit should be positive.
+	if quantumsToDeposit.Cmp(big.NewInt(0)) <= 0 {
+		return types.ErrInvalidDepositAmount
+	}
+	totalShares, exists := k.GetTotalShares(ctx, vaultId)
+	existingTotalShares := totalShares.NumShares.BigInt()
+	var sharesToMint *big.Int
+	if !exists || existingTotalShares.Cmp(big.NewInt(0)) < 1 {
+		// Mint `quoteQuantums` number of shares.
+		sharesToMint = quantumsToDeposit
+		// Initialize existingTotalShares as 0.
+		existingTotalShares = big.NewInt(0)
+	} else {
+		// Get vault equity.
+		equity, err := k.GetVaultEquity(ctx, vaultId)
+		if err != nil {
+			return err
+		}
+		// Don't mint shares if equity is non-positive.
+		if equity.Cmp(big.NewInt(0)) <= 0 {
+			return types.ErrNonPositiveEquity
+		}
+		// Mint `deposit (in quote quantums) * existing shares / vault equity (in quote quantums)`
+		// number of shares.
+		// For example:
+		// - a vault currently has 5000 shares and 4000 equity (in quote quantums)
+		// - each quote quantum is worth 5000 / 4000 = 1.25 shares
+		// - a deposit of 1000 quote quantums should thus be given 1000 * 1.25 = 1250 shares
+		// For shares calculation to be accurate, scale everything by denominator and mint
+		// numerator number of shares.
+		// For example:
```

Comment:
How do you expect this to work with owner shares being calculated? Each time you "scale" shares, each owner's total number of shares changes, wouldn't this lead to # state changes being linearly scaling with number of depositors for a vault when minting (undesirable) vs constant number of state changes per deposit?

Reply:
If each deposit happened with uneven quote quantums / equity, the total number of shares would continue scaling up and the integer would get larger and larger exponentially (imagine I always deposit quote quantums that is co-prime with the equity) right?

Reply:
So each time `quote quantums * existing shares` / `equity` doesn't lead to a denominator of 1 we scale the total number of shares up?
E.g. equity is now 1001, with total shares 1001. Let's say equity now increases to 1002. If I deposit 1 quote quantum again, the math is:
(1 * (1001)) / 1002 = 1001/1002. Now we scale the # of shares again up by 1002?

Reply:
conclusion from offline discussion: store shares as rational numbers